RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1470 @@
+              float const valueP_row = pmpcol[row];  //pulseMatrixP(j, ipulseReal);
+              float const value_row = pmcol[row];    //pulseMatrix(j, ipulseReal);
+              float const valueM_row = pmmcol[row];  //pulseMatrixM(j, ipulseReal);
+
+              float tmpprow = valueP_row - value_row;
+              float tmpmrow = valueM_row - value_row;
+
+              auto const covValue = 0.5 * (tmppcol * tmpprow + tmpmcol * tmpmrow);
+
+              covarianceMatrix(row, col) += ampl2 * covValue;
+            }
+          }
+        }
+      }
+
+      template <int NSAMPLES, int NPULSES>
+      class Kernel_minimize {
+      public:
+        template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+        ALPAKA_FN_ACC void operator()(TAcc const& acc,
+                                      OProductType::View outputGPU,
+                                      float const* amplitudes,
+                                      float* pulseMatrices,
+                                      float* pulseMatricesM,
+                                      float* pulseMatricesP,
+                                      HcalMahiPulseOffsetsPortableDevice::ConstView pulseOffsetsView,
+                                      float* noiseTerms,
+                                      float* electronicNoiseTerms,
+                                      int8_t* soiSamples,
+                                      HcalMahiConditionsPortableDevice::ConstView mahi,
+                                      bool const useEffectivePedestals,
+                                      IProductTypef01::ConstView f01HEDigis,
+                                      IProductTypef5::ConstView f5HBDigis,
+                                      IProductTypef3::ConstView f3HBDigis) const {
+          // can be relaxed if needed - minor updates are needed in that case!
+          static_assert(NPULSES == NSAMPLES);
+
+          auto const nchannels = f01HEDigis.size() + f5HBDigis.size() + f3HBDigis.size();
+
+          auto const nchannelsf015 = f01HEDigis.size() + f5HBDigis.size();
+
+          auto const threadsPerBlock(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[0u]);
+
+          //Loop over all groups of channels
+          for (auto group : uniform_groups_x(acc, nchannels)) {
+            //Loop over each channel
+            for (auto channel : uniform_group_elements_x(acc, group, nchannels)) {
+              auto const gch = channel.global;
+              auto const lch = channel.local;
+
+              if (gch >= nchannels)
+                return;
+
+              // if chi2 is set to -9999 do not run minimization
+              if (outputGPU.chi2()[gch] == -9999.f)
+                return;
+
+              // configure shared mem
+              float* shrmem = alpaka::getDynSharedMem<float>(acc);
+              float* shrMatrixLFnnlsStorage = shrmem + calo::multifit::MapSymM<float, NPULSES>::total * lch;
```

Comment:
OK, I'm confused :-(

The "shared memory" is actually different for each thread ?

Reply:
Could you add a comment to remind the reader about it ?

Reply:
If each thread uses its own portion of the shared memory, that would raise the question if the shared memory is needed in the first place, or if it could be replaced with a the matrix in function stack? (I could imagine there could be some tradeoffs here, that might be better to be studied after this PR)

Reply:
I think this approach comes from using shared memory as a kind of L1 cache, which was widespread in older GPU generations.

Probably to be revised for T4/L4 and newer GPUs, but definitely in a separate PR.